### ISSUE TITLE
`xm.save()` needs to synchronize functionalized tensors before calling `_xla_get_cpu_tensors()`

### ISSUE THREAD

**Issue by mcuiaws** [pytorch/xla#8482]
## 🐛 Bug

When xla functionalization is enabled, multiple live XlaTensors may be aliases of each other. A call to `SyncTensorsGraph()` can change the one one alias but not the other, breaking the aliasing relationship temporarily. The aliasing relationship is usually restored on any subsequent tensor operation from the torch side. However, `xm.save()` calls `_xla_get_cpu_tensors()` immediately after sync'ing the tensor graph, and incorrect values could be transferred.

## To Reproduce

Here's a small test case that shows wrong value being returned by `_xla_get_cpu_tensors()`:

```
import os

os.environ["XLA_DISABLE_FUNCTIONALIZATION"] = "0"
os.environ["XLA_ENABLE_PARAM_ALIASING"] = "0"

import torch
import torch_xla
import torch_xla.core.xla_model as xm

xla_to_cpu = torch_xla._XLAC._xla_get_cpu_tensors

def main():
    device = xm.xla_device()
    t1 = torch.tensor([1], device=device)
    t2 = t1.detach()
    xm.mark_step()

    t2.add_(t2)
    xm.mark_step()

    # mark_step() causes t1 and t2 to be out of sync on the XLA side.
    # _functionalize_sync() is needed to get them back in sync.
    # torch._sync(t1)

    t1_cpu, t2_cpu = xla_to_cpu([t1, t2])
    assert t1.item() == t1_cpu.item(), f"{t1.item()} vs {t1_cpu.item()}"

if __name__ == "__main__":
    main()
```

In this example I explicitly disabled `XLA_ENABLE_PARAM_ALIASING` to show that the wrong value is being returned. In our PJRT implementation, donated PjRtBuffers are not allowed to be read, and a CHECK failure occurs in `PjRtComputationClient::TransferFromDevice()`

## Expected behavior

The program should not assert.

## Environment

 - Reproducible on XLA backend [CPU/TPU/CUDA]: neuron trn1
 - torch_xla version: 2.5.1

**Comment by yaochengji** [comment]
Hi @mcuiaws 

Thanks for submitting the issue. I also submitted a similar PR to fix the DDP issue. https://github.com/pytorch/xla/pull/8489

An intriguing observation is that your test can now pass following my modifications. I suspect both issues stem from a lack of sync of functional tensors.

**Comment by mcuiaws** [comment]
@yaochengji Awesome! can you add the test case from #8483 to your PR?

**Comment by yaochengji** [comment]
It's added. cc @qihqi @tengyifei

**Comment by mcuiaws** [comment]
Fixed by #8489
